### PR TITLE
Borrowck: refactor visited map to a bitset

### PR DESCRIPTION
**Pull request by bugadani**
This PR refactors `Borrows`  and the `precompute_borrows_out_of_scope` function so that this initial phase has a much reduced memory pressure. This is achieved by reducing what is stored on the heap, and also reusing heap memory as much as possible.

### CODE REVIEW

**Comment by bugadani** [comment]
That map can actually be a set, so I'll have more changes on this.

**Comment by bjorn3** [comment]
The difference is indistinguishable from noise.

File: compiler/rustc_mir/src/borrow_check/mod.rs [hunk 1]
```diff
@@ -259,7 +259,7 @@ fn do_mir_borrowck<'a, 'tcx>(
 
     let regioncx = Rc::new(regioncx);
 
-    let flow_borrows = Borrows::new(tcx, &body, regioncx.clone(), &borrow_set)
+    let flow_borrows = Borrows::new(tcx, &body, regioncx.clone(), borrow_set.clone())
```
[the remark this answers is not in the document]
Comment (bugadani) [reply compiler/rustc_mir/src/borrow_check/mod.rs]: Actually, a plain old reference works as well.

**Comment by bugadani** [comment]
@matthewjasper sorry I snuck in an after-last-minute change, `Borrows` doesn't need to take `RC`s at all.

File: compiler/rustc_mir/src/dataflow/impls/borrows.rs [hunk 2]
```diff
@@ -30,132 +29,136 @@ pub struct Borrows<'a, 'tcx> {
...
     crate fn new(
         tcx: TyCtxt<'tcx>,
         body: &'a Body<'tcx>,
-        nonlexical_regioncx: Rc<RegionInferenceContext<'tcx>>,
-        borrow_set: &Rc<BorrowSet<'tcx>>,
+        nonlexical_regioncx: &'a RegionInferenceContext<'tcx>,
+        borrow_set: &'a BorrowSet<'tcx>,
     ) -> Self {
-        let mut borrows_out_of_scope_at_location = FxHashMap::default();
+        let mut prec = OutOfScopePrecomputer::new(body, nonlexical_regioncx.clone());
```
[the remark this answers is not in the document]
Comment (bugadani) [reply compiler/rustc_mir/src/dataflow/impls/borrows.rs]: OMG thanks for noticing!

**Pull request merged**